### PR TITLE
AGFS grad fee calc

### PR DESCRIPTION
**Pull request by jsugarman**
#### What
AGFS graduated fee calculation implementation for basic (basic) fee

#### Ticket

[CBO-298](https://dsdmoj.atlassian.net/browse/CBO-298)

#### Why
**As an** AGFS provider
**I want** the graduated basic (basic) fee that I claim to be calculated for me automatically on the graduated fee page
**So that** I do not have to spend time working this out myself and so that there is one source of truth for fees

#### How
- Use GraduatedPrice js and backend for basic [basic] fee with dummy unit values and modifiers where necessary to not account for the extra and additional fees. 
- supply different required values for cracked, retrial and cracked before retrial type

--------

#### TODO (wip)

 - [X] basic [basic] fee- BABAF
 - [X] trial, guilty plea, discontinuance 
 - [X] cracked trial
 - [X] retrial
 - [X] cracked before retrial

### CODE REVIEW

File: config/settings.yml [hunk 1]
```diff
@@ -88,9 +88,8 @@ active_features: []
 
 reject_refuse_messaging_released_at: <%= DateTime.new(2018, 4, 24, 23, 05, 0).change(offset: '+100') %>
 
-# agfs_fee_reform_release_date: <%= Date.new(2018, 4, 1) %>
-agfs_fee_reform_release_date: <%= ENV["ENV"].in?(['api-sandbox', 'demo', 'development']) ? Date.new(2018, 1, 1) : Date.new(2018, 04, 1) %>
-agfs_scheme_11_release_date: <%= ENV["ENV"].in?(['gamma', 'staging', 'disaster']) ? Date.new(2018, 12, 31) : Date.new(2018, 12, 1) %>
+agfs_fee_reform_release_date: <%= Date.new(2018, 4, 1) %>
+agfs_scheme_11_release_date: <%= Date.new(2018, 12, 31) %>
```
Comment (colinbruce) [inline config/settings.yml:92]: Generally :+1: for this, but should we let the software vendors know?

File: features/page_objects/sections/basic_fee_section.rb [hunk 2]
```diff
@@ -4,29 +4,49 @@
...
-    section :plea_and_trial_preparation_hearing, FeeSection, ".basic-fee-group.plea-and-trial-preparation-hearing"
-    section :conferences_and_views, FeeSection, ".basic-fee-group.conferences-and-views"
-    section :number_of_defendants_uplift, FeeSection, ".basic-fee-group.number-of-defendants-uplift"
-    section :number_of_cases_uplift, FeeCaseNumbersSection, ".basic-fee-group.number-of-cases-uplift"
-
-    sections :basic_fees_checklist, '#basic-fees .multiple-choice' do
-      element :label, 'label'
-      element :input, 'input'
+  section :basic_fee, FeeSection, ".basic-fee.fee-details"
+  # section :basic_fee_dates, FeeDatesSection, ".basic-fee.fee-details .fee-dates-row" # needed?
```
Comment (colinbruce) [inline features/page_objects/sections/basic_fee_section.rb]: Can be deleted ✂️ ?

File: features/support/tag_helper.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  def tag?(tag)
+    scenario_tags.include?(tag)
+  end
+
+  def fee_calc_vcr_tag?
+    tag?('@fee_calc_vcr')
+  end
+end
+
+World(TagHelper)
```
Comment (colinbruce) [inline features/support/tag_helper.rb]: No newline at file end

File: features/step_definitions/common_steps.rb [hunk 4]
```diff
@@ -172,7 +172,8 @@
 #
 And(/^I insert the VCR cassette '(.*?)'(?: and record '(.*?)')?$/) do |name, record|
   # Caching.clear if record
-  record_mode = (record || 'once').to_sym
+  record_mode = ENV['FEE_CALC_VCR_MODE']&.to_sym if fee_calc_vcr_tag?
+  record_mode ||= (record || 'once').to_sym
```
Comment (colinbruce) [inline features/step_definitions/common_steps.rb:176]: 🏆

**Review by colinbruce — APPROVED** [review]
Looking good. :+1:

File: app/assets/javascripts/modules/external_users/claims/fee_calculator/FeeCalculator.GraduatedPrice.js [hunk 5]
```diff
@@ -98,25 +108,29 @@
...
+        self.displayHelp(context, false);
         self.enableAmount(context);
       });
     },
 
-    buildFeeData: function(data, context) {
+    buildFeeData: function(data) {
       data.claim_id = $('#claim-form').data('claimId');
       data.price_type = 'GraduatedPrice';
+      var advocate_category = $('input:radio[name="claim[advocate_category]"]:checked').val();
```
Comment (jsugarman) [inline app/assets/javascripts/modules/external_users/claims/fee_calculator/FeeCalculator.GraduatedPrice.js]: refactor?

File: app/services/claims/fee_calculator/graduated_price.rb [hunk 6]
```diff
@@ -10,42 +10,111 @@ class GraduatedPrice < CalculatePrice
...
+      # TODO: warrant fees to be excluded until
       # - fee calculator amended to have codes for warrant fee scenarios
       # - CCCD is able to apply the sub category of warrant fee scenario logic
       #
-      def exclude_invalid_fee_calc
-        return unless %w[INWAR].include?(fee_type.unique_code)
-        raise StandardError, 'temporary exclusion of interim fee calc warrant fee types'
+      # TODO: retrial basic (basic) fee to excluded until
+      # - we confirm how to handle situations where retrial started before tria concluded
+      #
```
Comment (jsugarman) [inline app/services/claims/fee_calculator/graduated_price.rb:35]: add todo for cracked before retrial

File: config/settings.yml (hunk 1 above)
Comment (jsugarman) [reply config/settings.yml:92]: will do. 👍

File: features/page_objects/sections/basic_fee_section.rb (hunk 2 above)
Comment (jsugarman) [reply features/page_objects/sections/basic_fee_section.rb]: 👍

File: features/page_objects/sections/basic_fee_section.rb [hunk 7]
```diff
@@ -4,29 +4,49 @@
...
+
+  def checklist_item_for(label)
+    checklist.find do |item|
+      item.label.text.match?(Regexp.new(Regexp.escape(label), true))
     end
+  end
 
-    def checklist_labels
-      basic_fees_checklist.map { |item| item.label.text if item.has_label? }
+  def toggle(label)
```
Comment (jsugarman) [inline features/page_objects/sections/basic_fee_section.rb:38]: remove, unless needed

File: spec/factories/case_types.rb [hunk 8]
```diff
@@ -27,6 +27,22 @@
     roles                       %w{ agfs lgfs }
     uuid SecureRandom.uuid
 
+    trait :agfs_roles do
+      roles %w[agfs]
+    end
+
+    trait :lgfs_roles do
+      roles %w[agfs]
```
Comment (jsugarman) [inline spec/factories/case_types.rb]: should be %w[lgfs]

File: spec/services/claims/fee_calculator/graduated_price_spec.rb [hunk 9]
```diff
@@ -161,11 +149,237 @@
...
+              context 'with retrial interval within one calendar month' do
+                before do
+                  trial_end = 3.months.ago.to_date
+                  retrial_cracked = trial_end + 1.month
+                  allow(claim).to receive(:trial_cracked_at_third).and_return 'final_third'
+                  allow(claim).to receive(:trial_cracked_at).and_return retrial_cracked
+                end
+
+                context 'with retrial reduction requested' do
+                  # TODO: CCCD needs to offer "retrial reduction" option to user?? (as for retrials)
```
Comment (jsugarman) [inline spec/services/claims/fee_calculator/graduated_price_spec.rb]: move todo up or out

File: spec/services/claims/fee_calculator/graduated_price_spec.rb [hunk 10]
```diff
@@ -161,11 +149,237 @@
...
+              context 'with retrial interval greater than one calendar month' do
+                before do
+                  trial_end = 3.months.ago.to_date # TODO: no trial dates exist on cracked before retrial claims??
+                  retrial_cracked = trial_end + 1.month + 1.day
+                  allow(claim).to receive(:trial_cracked_at_third).and_return 'final_third'
+                  allow(claim).to receive(:trial_cracked_at).and_return retrial_cracked
+                end
+
+                context 'with retrial reduction requested' do
+                  # TODO: CCCD needs to offer "retrial reduction" option to user?? (as for retrials)
```
Comment (jsugarman) [inline spec/services/claims/fee_calculator/graduated_price_spec.rb]: move up or out

File: config/settings.yml (hunk 1 above)
Comment (jsugarman) [reply config/settings.yml:92]: vendors warned of this in #software-vendor-adp

**Pull request merged**